Improved Istio Getting Started (#6642)
istio/istio.io: content/en/docs/setup/getting-started/index.md

keywords: [getting-started, install, bookinfo, quick-start, kubernetes]
---

This guide lets you quickly evaluate Istio.

These steps assume you are using

> [Minikube](https://kubernetes.io/docs/tasks/tools/install-minikube/) and have a

Reviewer: Assuming minikube doesn't sound like a simplification. Minikube is not the quickest way to get started unless minikube is already installed. Creating a cluster using Docker Desktop or IBM Cloud (and I assume other clouds like GCE) is an easier way to start from scratch than installing and setting up minikube and minikube tunnel.

These instructions should just say you need to start with a Kubernetes cluster (it can mention minikube as one way to do it), but assuming minikube throughout the instructions is not a good idea IMO.

Reviewer 2: I don't know that IBM Cloud is unless they also have an account and the CLI installed. Similar for GKE and others. I'm not sure if they would already have Docker installed (we ask our developers to use it, but potential uses kicking the tires?).

Reviewer: @ericvn Yes, that sounds good. In fact, that's similar to what linkerd does in their quick start. They just say that you need a kubernetes cluster and they point to minikube and docker desktop as a couple of easy choices if you're also new to Kubernetes.

Regarding IBM Cloud or Docker Desktop being easier than minikube, I think that a good percentage of the people that are looking at Istio for the first time are not also new to Kubernetes and Docker. I'm assuming that many of the people trying to get started with Istio already use Kubernetes, and/or Docker, so they probably already have something like IBM Cloud or Docker Desktop or Minikube configured. So we can't say that regardless of what may be the easiest way to set up a cluster in you case, these instructions are specifically for minikube, so if you want to get started quickly with Istio you need to install minikube first.

Author: Users were failing to successfully complete the old Getting Started and the feedback was asking us to ensure we had the exact successful steps documented for a particular install path, minikube being the one that's free and accessible to everyone as our common starting point.

But we add note at the beginning the platform setup link and provide the "Other" tab at the mid-point so someone can still be successful using IBM or GKE if they prefer. Maybe we should just rename the Other tab to be more specific about IBM cloud, GKE, etc?

Reviewer: I see, so the instructions don't really "assume you are using minikube". The instructions work for any platform, it's just that minikube is the fast path. I guess that's OK, but you should reword the above sentence to just say that the user needs a Kubernetes cluster to get started, and mention minikube as a good choice for somebody also new to k8s. Then you can make minikube the simplest option in the following instructions, just so long as the instructions also work well on the other platforms too

Author: fixed

{{< gloss >}}cluster{{< /gloss >}} running a compatible version of Kubernetes.
Minikube is a tool that makes it easy to run Kubernetes locally with a
single-node Kubernetes cluster inside a Virtual Machine (VM) on your computer.

To use a cluster on a different platform or in the cloud, follow the appropriate
[platform-specific setup instructions](/docs/setup/platform-setup/) before you
continue.

Follow these steps to get started with Istio:

> 1. [Download and install Istio](#download)

Reviewer 2: You have two sections, so not sure if combining the into one is beneficial or not. I feel breaking this into two items and pointing to each section makes the most sense.

Author: fixed - they are broken up in the content, but for this summary at the beginning we mention. them combined.

1. [Deploy the sample application](#bookinfo)
1. [Open the application to outside traffic](#ip)
1. [View the dashboard](#dashboard)

> ## Install Istio {#download}

Reviewer: This section (#download) is referenced from all the other more advanced install instructions, so you can't change it to do install as well.

Reviewer: ```suggestion
## Download Istio {#download}
```

Author: fixed

    $ cd istio-{{< istio_full_version >}}
    {{< /text >}}

    The installation directory contains:

    - Installation YAML files for Kubernetes in `install/kubernetes`
    - Sample applications in `samples/`
    - The [`istioctl`](/docs/reference/commands/istioctl) client binary in the

>       `bin/` directory. `istioctl` is used if you choose to manually inject

Reviewer 2: `istioctl` is also used to get at the dashboard and install Istio. I would be inclined to simply remove the `istioctl is used ...` instead of adding a list of items.

Author: fixed

    $ cd istio-{{< istio_full_version >}}
    {{< /text >}}

1.  Add the `istioctl` client to your path (Linux or macOS):

> 

Reviewer: Like I said above, these instructions are pointed to by all the more advanced install options, so this is pretty useful information that we probably don't want to lose.

Author: fixed

    {{< text bash >}}
    $ export PATH=$PWD/bin:$PATH
    {{< /text >}}

> 1.  For this tutorial, we use the `demo` profile. It's selected to have a good set

Reviewer: This isn't a "tutorial", it's the setup instructions.

Author: fixed

Reviewer: Also should we link to the configure profile page?
```suggestion
1.  For this tutorial, we use the `demo` [configuration profile](link). It's selected to have a good set
```

Author: fixed

    of defaults for testing, but there are other profiles for production or

>     performance testing.

Author: fixed

    {{< text bash >}}
    $ istioctl dashboard kiali
    {{< /text >}}

1.  In the left navigation menu, select _Graph_ and in the _Namespace_ drop down, select _default_.

    The Kiali dashboard shows an overview of your mesh with the relationships
    between the services in the `Bookinfo` sample application. It also provides
    filters to visualize the traffic flow.

    {{< image link="./kiali-example2.png" caption="Kiali Dashboard" >}}

## Next steps

Congratulations on completing the installation!

These tasks are a great place for beginners to start using Istio features:

- [Request routing](/docs/tasks/traffic-management/request-routing/)
- [Fault injection](/docs/tasks/traffic-management/fault-injection/)
- [Traffic shifting](/docs/tasks/traffic-management/traffic-shifting/)
- [Querying metrics](/docs/tasks/observability/metrics/querying-metrics/)
- [Visualizing metrics](/docs/tasks/observability/metrics/using-istio-dashboard/)
- [Rate limiting](/docs/tasks/policy-enforcement/rate-limiting/)
- [Accessing external services](/docs/tasks/traffic-management/egress/egress-control/)
- [Visualizing your mesh](/docs/tasks/observability/kiali/)

> Before you customize Istio to fit your platform and intended use,

Reviewer 3: This content seems out of place in getting started. I wonder if it should be promoted to its own page.

Reviewer 3: It feels like "what is Istio". Ok, you installed Istio, ready to promote to prod?

Author: Reworded this section to be more clear about what we're telling the user to do.

    productpage-v1-85b9bf9cd7-zg7tr   0/2     PodInitializing   0          2m29s
    ratings-v1-6c9dbf6b45-5djtx       0/2     PodInitializing   0          2m29s
    reviews-v1-564b97f875-dzdt5       0/2     PodInitializing   0          2m30s
    reviews-v2-568c7c9d8f-p5wrj       1/2     Running           0          2m29s
    reviews-v3-67b4988599-7nhwz       0/2     PodInitializing   0          2m29s
    {{< /text >}}

    {{< tip >}}
    Re-run the previous command and wait until all pods report READY 2 / 2 and
    STATUS Running before you go to the next step. This might take a few minutes
    depending on your platform.
    {{< /tip >}}

1.  Verify everything is working correctly up to this point. Run this command to
    see if the app is running inside the cluster and serving HTML pages by
    checking for the page title in the response:

    {{< text bash >}}
    $ kubectl exec -it $(kubectl get pod -l app=ratings -o jsonpath='{.items[0].metadata.name}') -c ratings -- curl productpage:9080/productpage | grep -o "<title>.*</title>"
    <title>Simple Bookstore App</title>
    {{< /text >}}

## Open the application to outside traffic {#ip}

> Your app is deployed but not accessible from the outside. To make it accessible,

Reviewer: ```suggestion
The Bookinfo application is deployed but not accessible from the outside. To make it accessible,
```

you need to create an
[Istio Ingress Gateway](/docs/concepts/traffic-management/#gateways), which maps a path to a
route at the edge of your mesh.

> 1.  Connect this application to be published by the Istio gateway:

Reviewer 2: Grammer..  c/by/to might be enough to make it sound OK.

Author: fixed

    $ kubectl get gateway
    NAME               AGE
    bookinfo-gateway   32s
    {{< /text >}}

### Determining the ingress IP and ports

Follow these instructions to set the `INGRESS_HOST` and `INGRESS_PORT` variables
for accessing the gateway. Use the tabs to choose the instructions for your
chosen platform:

{{< tabset category-name="gateway-ip" >}}

{{< tab name="Minikube" category-value="external-lb" >}}

Set the ingress ports:

{{< text bash >}}
$ export INGRESS_PORT=$(kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="http2")].nodePort}')
$ export SECURE_INGRESS_PORT=$(kubectl -n istio-system get service istio-ingressgateway -o jsonpath='{.spec.ports[?(@.name=="https")].nodePort}')
{{< /text >}}

Ensure a port was successfully assigned to each environment variable:

{{< text bash >}}
$ echo $INGRESS_PORT
32194
{{< /text >}}

{{< text bash >}}
$ echo $SECURE_INGRESS_PORT
31632
{{< /text >}}

Set the ingress IP:

{{< text bash >}}
$ export INGRESS_HOST=$(minikube ip)
{{< /text >}}

Ensure an IP address was successfully assigned to the environment variable:

{{< text bash >}}
$ echo $INGRESS_HOST
192.168.4.102
{{< /text >}}

> Run this command in a new terminal window to start a Minikube tunnel that

Reviewer 2: I'm still not sure this, the `minikube tunnel` is needed. It probably doesn't hurt.

Author: I had to add it for my environment to work. It literally was dead for hours until specifically this step.

{{< /text >}}

You need to create firewall rules to allow the TCP traffic to the `ingressgateway` service's ports.
Run the following commands to allow the traffic for the HTTP port, the secure port (HTTPS) or both:

{{< text bash >}}
$ gcloud compute firewall-rules create allow-gateway-http --allow tcp:$INGRESS_PORT
$ gcloud compute firewall-rules create allow-gateway-https --allow tcp:$SECURE_INGRESS_PORT
{{< /text >}}

_Docker For Desktop:_

{{< text bash >}}
$ export INGRESS_HOST=127.0.0.1
{{< /text >}}

_Other environments (e.g., IBM Cloud Private, etc.):_

{{< text bash >}}
$ export INGRESS_HOST=$(kubectl get po -l istio=ingressgateway -n istio-system -o jsonpath='{.items[0].status.hostIP}')
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

1.  Set `GATEWAY_URL`:

    {{< text bash >}}
    $ export GATEWAY_URL=$INGRESS_HOST:$INGRESS_PORT
    {{< /text >}}

1.  Ensure an IP address and port were successfully assigned to the environment variable:

    {{< text bash >}}
    $ echo $GATEWAY_URL
    192.168.99.100:32194
    {{< /text >}}

### Verify external access {#confirm}

Confirm that the Bookinfo application is accessible from outside the cluster by

> navigating in your browser to `http://<your $GATEWAY_URL value>/productpage`.

Reviewer 2: Put this command in a text bash.

Author: This isn't a bash command. But is there a way to open the GUI browser to a specific URL using a bash command?

Reviewer 2: I think you can just use text. I incorrectly had bash in my prior remark.

Author: fixed

## View the dashboard {#dashboard}

> Istio has several optional dashboards installed by the `demo` installation. The

Reviewer 2: Is this saying the dashboards are installed by the installation?  I 'm not sure that's 100% accurate. On the flip side, if I install Kiali outside of Istio, can I still get at it's dashboard with the CLI? @esnible, do you know?

Author: The demo install definitely adds the Kiali dashboard per my testing. The others I'm not sure.

Run the following commands to allow the traffic for the HTTP port, the secure port (HTTPS) or both:

{{< text bash >}}
$ gcloud compute firewall-rules create allow-gateway-http --allow tcp:$INGRESS_PORT
$ gcloud compute firewall-rules create allow-gateway-https --allow tcp:$SECURE_INGRESS_PORT
{{< /text >}}

_Docker For Desktop:_

{{< text bash >}}
$ export INGRESS_HOST=127.0.0.1
{{< /text >}}

_Other environments (e.g., IBM Cloud Private, etc.):_

{{< text bash >}}
$ export INGRESS_HOST=$(kubectl get po -l istio=ingressgateway -n istio-system -o jsonpath='{.items[0].status.hostIP}')
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

1.  Set `GATEWAY_URL`:

    {{< text bash >}}
    $ export GATEWAY_URL=$INGRESS_HOST:$INGRESS_PORT
    {{< /text >}}

1.  Ensure an IP address and port were successfully assigned to the environment variable:

    {{< text bash >}}
    $ echo $GATEWAY_URL
    192.168.99.100:32194
    {{< /text >}}

### Verify external access {#confirm}

Confirm that the Bookinfo application is accessible from outside the cluster by
navigating in your browser to:

{{< text plain >}}

> `http://<your $GATEWAY_URL value>/productpage`

Reviewer: This URL needs to be wrapped in back-ticks or a text section, but not both. Pick one or the other. Personally, I think back-ticks are better since the features of a text section (e.g., copy button) aren't really that useful here, since most of the string (`<your $GATEWAY_URL value>`) needs to be replaced before using it anyway.

Reviewer 2: And I had previously asked him to put in a text box, so I'll let Adam decide on which way to go with this. I also did see in the Learning series of pages, where they have an _echo_ in a text bash box to actually output the URL (it'll expand the envvars), and have you cut/paste from that output to the browser.

Reviewer: Tough call whether or not the echo approach is too basic-tutorial like for the setup instructions, but it is a simple way to get the right value for the browser.

This does get me to wonder if starting the bookinfo app in the setup getting started page is messing up the flow for other docs. The bookinfo example points to the setup instructions as a before you begin step, and then proceeds to give the instructions to start the sample. If the user followed the get started setup instructions, instead of other more advanced install instructions, the following bookinfo instructions are not needed anymore, since bookinfo is already up and running. I think get started and the bookinfo example need to add some text to explain that the get started install instructions are a quick start including bookinfo, so the bookinfo sample instructions are only for users that didn't use the quick start. Otherwise people will be confused about what instructions they need to follow.

Reviewer 2: I have that same concern and noted it somewhere earlier (can't find it right now). We point to getting started in some docs and I don't think they expect that Bookinfo is necessarily installed after that is complete.

Author: Going with the echo approach as it is the most user friendly and most accurate. 

Regarding the other comments, the Getting Started is getting hard to optimize for its original purpose because so many other docs depend on it for its sub-sections, e.g. other docs linking to its downloading Istio section, and others. Also when another doc links to Getting Started for one of its sub-sections, readers can easily make the mistake of forgetting to go back to the original doc at the immediate conclusion of the given subsection.

A better way to handle this situation is to remake the truly shareable subsections as single-sourced boilerplates instead of so much linking back and forth. This lets you use the instructions in context everywhere it's needed without linking so much, maintaining single-sourced content re-use, but also letting you remove the boilerplate and customize something one-off as needed in the places where the info truly needs to be different.

Reviewer: @adammil2000 I don't think there's really that much linking back and forth with getting started (downloading section is the main one I know about). The bigger problem is that lots of docs link generally back to the setup docs, not getting started specifically, to tell people they need to install istio before they can do pretty much anything else. Users can choose from several ways to install istio, but getting started is the doc that stands out as the doc in the setup section to click first. So, I think that what's missing in this new version is a short description at the top of the doc, similar to what was there before, that says that this is a quick start for new users, but if you are already familiar with Istio or interested in installing other configuration profiles or a more advanced deployment model, follow the installing with istioctl instructions instead. The bookinfo sample should also have a tip/warning in the before you begin section that says if you've installed Istio with the getting started (quick start) guide, you will already have bookinfo deployed and can ignore the following instructions.

Author: fixed

    deploy along with it.

    {{< text bash >}}
    $ kubectl get services
    NAME          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
    details       ClusterIP   10.0.0.212      <none>        9080/TCP   29s
    kubernetes    ClusterIP   10.0.0.1        <none>        443/TCP    25m
    productpage   ClusterIP   10.0.0.57       <none>        9080/TCP   28s
    ratings       ClusterIP   10.0.0.33       <none>        9080/TCP   29s
    reviews       ClusterIP   10.0.0.28       <none>        9080/TCP   29s
    {{< /text >}}

    and

    {{< text bash >}}
    $ kubectl get pods
    NAME                              READY   STATUS            RESTARTS   AGE
    details-v1-78d78fbddf-tj56d       0/2     PodInitializing   0          2m30s
    productpage-v1-85b9bf9cd7-zg7tr   0/2     PodInitializing   0          2m29s
    ratings-v1-6c9dbf6b45-5djtx       0/2     PodInitializing   0          2m29s
    reviews-v1-564b97f875-dzdt5       0/2     PodInitializing   0          2m30s
    reviews-v2-568c7c9d8f-p5wrj       1/2     Running           0          2m29s
    reviews-v3-67b4988599-7nhwz       0/2     PodInitializing   0          2m29s
    {{< /text >}}

    {{< tip >}}

>     Re-run the previous command and wait until all pods report READY 2 / 2 and

Reviewer 4: @howardjohn had a good example in another repo. This might work instead:

kubectl wait --for=condition=Available deployment --all --timeout=120s -n default

Run the following commands to allow the traffic for the HTTP port, the secure port (HTTPS) or both:

{{< text bash >}}
$ gcloud compute firewall-rules create allow-gateway-http --allow tcp:$INGRESS_PORT
$ gcloud compute firewall-rules create allow-gateway-https --allow tcp:$SECURE_INGRESS_PORT
{{< /text >}}

_Docker For Desktop:_

{{< text bash >}}
$ export INGRESS_HOST=127.0.0.1
{{< /text >}}

_Other environments (e.g., IBM Cloud Private, etc.):_

{{< text bash >}}
$ export INGRESS_HOST=$(kubectl get po -l istio=ingressgateway -n istio-system -o jsonpath='{.items[0].status.hostIP}')
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

1.  Set `GATEWAY_URL`:

    {{< text bash >}}
    $ export GATEWAY_URL=$INGRESS_HOST:$INGRESS_PORT
    {{< /text >}}

1.  Ensure an IP address and port were successfully assigned to the environment variable:

    {{< text bash >}}
    $ echo $GATEWAY_URL
    192.168.99.100:32194
    {{< /text >}}

### Verify external access {#confirm}

Confirm that the Bookinfo application is accessible from outside. Copy the
output of this command and paste into your browser:

{{< text bash >}}

> $ echo http://$GATEWAY_URL/productpage

Reviewer 5: Is this actually valid thing to put into browser?
`GATEWAY_URL` will have internal node port set (here: `32194`).
Yet output from:
```sh
kubectl get svc istio-ingressgateway -n istio-system
```
shows that this is **destination** port for request that will originate **to** `http://${INGRESS_HOST}:80` (!)
The former one (with port `80`) works for me while testing. The one with `nodePort` gives no responses at all - from local browser.

Reviewer 4: Hi pazkooda! Please raise this as an issue, rather than a comment on this merged PR, so it can be triaged and looked at.  Thanks!
